src/EditorFeatures/Core/Implementation/IntelliSense/EditorCompletionImplementation/CompletionItemSource.cs

```diff
@@ -0,0 +1,249 @@
+        }
+
+        public async Task<object> GetDescriptionAsync(IAsyncCompletionSession session, EditorCompletion.CompletionItem item, CancellationToken cancellationToken)
+        {
+            if (!item.Properties.TryGetProperty<RoslynCompletionItem>(RoslynItem, out var roslynItem) ||
+                !item.Properties.TryGetProperty<ITextBuffer>(TriggerBuffer, out var triggerBuffer))
+            {
+                return null;
+            }
+
+            var document = triggerBuffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges();
+            if (document == null)
+            {
+                return null;
+            }
+
+            var service = (CompletionServiceWithProviders)document.GetLanguageService<CompletionService>();
+            var description = await service.GetProvider(roslynItem).GetDescriptionAsync(document, roslynItem, cancellationToken).ConfigureAwait(false);
+
+            return new ClassifiedTextElement(description.TaggedParts.Select(p => new ClassifiedTextRun(p.Tag.ToClassificationTypeName(), p.Text)));
+        }
+
+        private static RoslynTrigger GetRoslynTrigger(EditorCompletion.CompletionTrigger trigger, SnapshotPoint triggerLocation)
+        {
+            switch (trigger.Reason)
+            {
+                case EditorCompletion.CompletionTriggerReason.Invoke:
+                case EditorCompletion.CompletionTriggerReason.InvokeAndCommitIfUnique:
+                    return RoslynTrigger.Invoke;
+                case EditorCompletion.CompletionTriggerReason.Insertion:
+                    return RoslynTrigger.CreateInsertionTrigger(trigger.Character);
+                case EditorCompletion.CompletionTriggerReason.Deletion:
+                    return RoslynTrigger.CreateDeletionTrigger(trigger.Character);
+                case EditorCompletion.CompletionTriggerReason.Backspace:
+                    var snapshotBeforeEdit = trigger.ViewSnapshotBeforeTrigger;
+                    char characterRemoved;
+                    if (triggerLocation.Position >= 0 && triggerLocation.Position < snapshotBeforeEdit.Length)
+                    {
+                        // If multiple characters were removed (selection), this finds the first character. 
+                        // Maybe it should be re-considered to find the last removed character.
+                        characterRemoved = snapshotBeforeEdit[triggerLocation.Position];
+                    }
+                    else
+                    {
+                        characterRemoved = (char)0;
+                    }
+
+                    return RoslynTrigger.CreateDeletionTrigger(characterRemoved);
+                case EditorCompletion.CompletionTriggerReason.SnippetsMode:
+                    return new RoslynTrigger(CompletionTriggerKind.Snippets);
+            }
+
+            throw ExceptionUtilities.UnexpectedValue(trigger.Reason);
```

Comment:
i don't like this.  We're switching over an enum not in our control.  If there's a case we don't know about, we should juts bail gracefully, not tear down hte stack.

Reply:
1. We could migrate to the VS trigger but this requires migrating a lot of code that will be removed when the new completion will start.

2. If we have a new scenario we should support and we do not know how to do so, it is better to be alerted rather than silently process it somehow.

I prefer to:
1. keep track of any new changes/scenarios that could happen with this hook.
2. consider removing the Roslyn trigger with removing the legacy completion

---
In reply to: [222083344](https://github.com/dotnet/roslyn/pull/29016#discussion_r222083344) [](ancestors = 222083344)

Reply:
> If we have a new scenario we should support and we do not know how to do so, it is better to be alerted rather than silently process it somehow.

You can NFW then.  You should not throw an exception.  This can either tear down the process, end up with goldbars, or other undesirable scenarios.  It simply is not a good coding pattern to throw exceptions whne switching over enums out of your control.